File: examples/python/example_callback.py
import clarabel
import numpy as np
from scipy import sparse


def termination_callback(info: clarabel.DefaultInfo) -> bool:
    """
    Evaluate solver info and return a boolean result.

    Parameters
    ----------
    info : clarabel.DefaultInfo
        The solver info object to evaluate

    Returns
    -------
    bool
        True to signal termination
    """
    if info.iterations < 3:
        print("tick")
        return False  # continue
    else:
        print("BOOM!")
        return True  # stop


P = sparse.csc_matrix([[4., 1.], [1., 2.]])
P = sparse.triu(P).tocsc()

A = sparse.csc_matrix([
    [-1., -1.],
    [-1.,  0.],
    [0.,  -1.],
    [1.,   1.],
    [1.,   0.],
    [0.,   1.]])

q = np.array([1., 1.])
b = np.array([-1., 0., 0., 1., 0.7, 0.7])

cones = [clarabel.NonnegativeConeT(3), clarabel.NonnegativeConeT(3)]
settings = clarabel.DefaultSettings()

# ------------------------------------------------------------
# Solve using a callback
# ------------------------------------------------------------

# solves
solver = clarabel.DefaultSolver(P, q, A, b, cones, settings)
solver.solve()

# stops early
solver.set_termination_callback(termination_callback)
solver.solve()

# works again
solver.unset_termination_callback()
solver.solve()

# ------------------------------------------------------------
# # Solve using a callback with state
# ------------------------------------------------------------

counter = [-1]  # Using a list as a mutable container


def termination_callback_with_state(info: clarabel.DefaultInfo) -> bool:
    # Increment the counter each time the callback is called
    counter[0] += 1
    print(f"Iteration {info.iterations}: Counter = {counter[0]}")

    # Terminate when counter reaches 3
    if counter[0] >= 3:
        return True
    return False


# solves
solver = clarabel.DefaultSolver(P, q, A, b, cones, settings)
solver.solve()

# stops early
solver.set_termination_callback(termination_callback_with_state)
solver.solve()

# works again
solver.unset_termination_callback()
solver.solve()

File: examples/rust/example_callback.rs
#![allow(non_snake_case)]
use clarabel::algebra::*;
use clarabel::solver::*;

fn main() {
    // This example shows how to use a custom termination function
    // to halt the solver

    let P: CscMatrix<f64> = CscMatrix::<f64>::zeros((2, 2));
    let q = vec![1., -1.];

    //a 2-d box constraint, separated into 4 inequalities.
    //A = [I; -I]
    let _A = CscMatrix::new(
        4,                      // m
        2,                      // n
        vec![0, 2, 4],          // colptr
        vec![0, 2, 1, 3],       // rowval
        vec![1., -1., 1., -1.], // nzval
    );

    // easier way - use the From trait to construct A:
    let A = CscMatrix::from(&[
        [1., 0.],  //
        [0., 1.],  //
        [-1., 0.], //
        [0., -1.], //
    ]);

    let b = vec![1.; 4];

    let cones = [NonnegativeConeT(4)];

    let settings = DefaultSettingsBuilder::default()
        .equilibrate_enable(true)
        .max_iter(50)
        .build()
        .unwrap();

    let mut solver = DefaultSolver::new(&P, &q, &A, &b, &cones, settings).unwrap();

    // setup a custom termination function
    let f = |info: &DefaultInfo<f64>| {
        if info.iterations < 3 {
            println!("tick");
            false //continue
        } else {
            println!("BOOM!\n");
            true // stop
        }
    };

    solver.set_termination_callback(f);
    solver.solve();

    // setup a custom termination function that carries internal state
    let mut counter = 0;
    let fmut = move |info: &DefaultInfo<f64>| {
        counter += 1;
        println!("Iteration {}: counter = {}", info.iterations, counter);
        if counter < 3 {
            false // continue
        } else {
            println!("BOOM!\n");
            true // stop
        }
    };

    solver.set_termination_callback(fmut);
    solver.solve();

    // turn it off and run again
    solver.unset_termination_callback();
    solver.solve();
}
